feat(amqp sink): add priority property

aramperes opened the pull request:
<!--
  Your PR title must conform to the conventional commit spec:
  https://www.conventionalcommits.org/en/v1.0.0/

  <type>(<scope>)!: <description>

  * `type` = chore, enhancement, feat, fix, docs, revert
  * `!` = OPTIONAL: signals a breaking change
  * `scope` = Optional when `type` is "chore" or "docs", available scopes https://github.com/vectordotdev/vector/blob/master/.github/workflows/semantic.yml#L31
  * `description` = short description of the change

Examples:

  * enhancement(file source): Add `sort` option to sort discovered files
  * feat(new source): Initial `statsd` source
  * fix(file source): Fix a bug discovering new files
  * chore(external docs): Clarify `batch_size` option
-->

## Summary
<!-- Please provide a brief summary about what this PR does.
This should help the reviewers give feedback faster and with higher quality. -->

Provides the option to set the `priority` of messages sent by the `amqp` sink. [RabbitMQ documentation](https://www.rabbitmq.com/docs/priority).

Priority is a useful way to order messages when they are sent to the same exchange/queue through different sinks.

The property is a numeric value (`u8`) that also allows templating. The template must resolve to an integer between `0-255`, otherwise the priority will not be set on the message (default behavior is priority `0`). An invalid template will log and drop the event.

## Change Type
- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?
- [ ] Yes
- [x] No

## How did you test this PR?

1. Declare a RabbitMQ queue with the `x-max-priority` argument set (values up to 5 recommended).
2. Declare an exchange, and bind the queue.
3. Configure an `amqp` sink in Vector, publishing to the `exchange` and with `properties.priority` configured. It can be a template: (`"{{ .priority }}"`), or a constant: `1`.

```yaml
sources:
  in:
    type: stdin

transforms:
  prioritize:
    type: remap
    inputs:
      - in
    source: |
      .priority = random_int(1, 10)

sinks:
  sink_amqp:
    type: amqp
    inputs:
      - prioritize
    connection_string: amqp://user:password@127.0.0.1:5672/%2f?timeout=10
    exchange: example
    encoding:
      codec: json
    properties:
      priority: "{{ .priority }}"
```

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist
- [x] Please read our [Vector contributor resources](https://github.com/vectordotdev/vector/tree/master/docs#getting-started).
  - `make check-all` is a good command to run locally. This check is
    defined [here](https://github.com/vectordotdev/vector/blob/1ef01aeeef592c21d32ba4d663e199f0608f615b/Makefile#L450-L454). Some of these
    checks might not be relevant to your PR. For Rust changes, at the very least you should run:
    - `cargo fmt --all`
    - `cargo clippy --workspace --all-targets -- -D warnings`
    - `cargo nextest run --workspace` (alternatively, you can run `cargo test --all`)
- [x] If this PR introduces changes Vector dependencies (modifies `Cargo.lock`), please
  run `dd-rust-license-tool write` to regenerate the [license inventory](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv) and commit the changes (if any). More details [here](https://crates.io/crates/dd-rust-license-tool).

## References

<!-- Please list any issues closed by this PR. -->

<!--
- Closes: <issue link>
-->

<!-- Any other issues or PRs relevant to this PR? Feel free to list them here. -->

-  [RabbitMQ documentation](https://www.rabbitmq.com/docs/priority).

aramperes commented on src/sinks/amqp/config.rs:
```diff
@@ -19,6 +19,9 @@ pub struct AmqpPropertiesConfig {
     pub(crate) expiration_ms: Option<u64>,
+
+    /// Priority for AMQP messages.
+    pub(crate) priority: Option<u8>,
```
Optimally, it would be possible to make this property a `Template` so it can be dynamically calculated per event. I could not find other examples where a `Template` is expected to output a `uint`. The `str` output of a `Template` could be parsed as a `uint`, but would be error-prone if the result is not numeric, or out-of-bounds.

aramperes commented:
Let me see if I can make Templating work here... It would prevent a future breaking change here

jszwedko commented:
> > Let me see if I can make Templating work here... It would prevent a future breaking change here
> 
> Sounds good. (Although, I don't think the change would be breaking since the config would accept both constants and templated values.)

Yeah, exactly. Templates can just be raw literals like `foo` so I think this _could_ be added after and wouldn't be breaking. It would be useful to be able to use templates for non-string literals for sure.

aramperes commented:
Fair enough; my thinking currently is to make it default to `0` (same as no priority) if the templating fails to produce a `u8`. That would be kinda consistent with RabbitMQ behaviour being permissive:

> Messages without a priority property are treated as if their priority were 0. Messages with a priority which is higher than the queue's maximum are treated as if they were published with the maximum priority.

Would that be better than a runtime error, in the event that the template renders an invalid type?

I'm working on a integration test (with templating) right now.

jszwedko commented:
> > Messages without a priority property are treated as if their priority were 0. Messages with a priority which is higher than the queue's maximum are treated as if they were published with the maximum priority.
> 
> Would that be better than a runtime error, in the event that the template renders an invalid type?

Good question 🤔 Typically for "template errors" we consider them fatal and drop the event, but I can see the argument here that there is a safe default. I'd be 👍 to falling back to that default and just emitting a warning if the template fails to render correctly (rather than an error).

aramperes commented:
I updated the configuration to allow templating of `properties.priority`. Example:

```yaml
sources:
  in:
    type: stdin

transforms:
  prioritize:
    type: remap
    inputs:
      - in
    source: |
      .priority = random_int(1, 10)

sinks:
  sink_amqp:
    type: amqp
    inputs:
      - prioritize
    connection_string: amqp://user:password@127.0.0.1:5672/%2f?timeout=10
    exchange: example
    encoding:
      codec: json
    properties:
      priority: "{{ priority }}"
```

Because it is treated as a string, it must be `"quoted"` in YAML configurations, which may be counter intuitive. The error thrown in the case `priority: 1` is:

```
Configuration error. error=invalid type: integer `1`, expected a string
in `sinks.sink_amqp`
```

I also added integration tests for the various cases. I did not cover the case where an invalid template drops the event; if you could point me to an example on how to catch dropped events through instrumentation/counters I could add a case for that.

pront commented:
First of all, thanks for extending this and including the tests 👍 

> Because it is treated as a string, it must be `"quoted"` in YAML configurations, which may be counter intuitive. The error thrown in the case `priority: 1` is:
> 
> ```
> Configuration error. error=invalid type: integer `1`, expected a string
> in `sinks.sink_amqp`
> ```
 
Hmm, this is not great UX, especially since the rabbitmq docs clearly ask for a number. Still if we extend templating to accept numbers as well it improve the UX automatically. Maybe worth adding a note to the docs so users aren't surprised cc @jszwedko

jszwedko commented:
> First of all, thanks for extending this and including the tests 👍
> 
> > Because it is treated as a string, it must be `"quoted"` in YAML configurations, which may be counter intuitive. The error thrown in the case `priority: 1` is:
> > ```
> > Configuration error. error=invalid type: integer `1`, expected a string
> > in `sinks.sink_amqp`
> > ```
> 
> Hmm, this is not great UX, especially since the rabbitmq docs clearly ask for a number. Still if we extend templating to accept numbers as well it improve the UX automatically. Maybe worth adding a note to the docs so users aren't surprised cc @jszwedko

Agreed, the UX here isn't great for literals. Could we change the way `Template` deserializes to allow numeric literals it addition to string literals? That would need to be a change to the `Template` struct (https://github.com/vectordotdev/vector/blob/21776bbd15113e34841dc6269358fc69b9166546/src/template.rs). You can already see there how it handles strings.

Again, happy to see the original PR go in without template support if you don't want to take that on right now 🙂

aramperes commented:
Agreed, I will give it a shot!

aramperes commented:
I pushed a solution for the configuration of numeric values as optional templates:

`Template` now deserializes from a new enum `TemplateSource`, which supports numeric values (`i64`, `u64`, `f64`) and `String`. The `TemplateSource` is then treated as a `String` and the template is parsed/rendered as before.

There could be a performance optimization done for numeric values to skip the `Number -> TemplateSource -> String -> Template -> String -> u8` cycle when evaluating the AMQP template, but that would require some more changes to the `Template` struct.

For serialization, it is simply converting to a `TemplateSource::String` regardless of the original type. I'm not clear if that causes any UX issues; otherwise `Template` could be updated to maintain `TemplateSource` information. That would be a bit more work to get done.

Tested with both configs:

```yaml
sinks:
  sink_amqp:
    type: amqp
    inputs:
      - prioritize
    connection_string: amqp://user:password@127.0.0.1:5672/%2f?timeout=10
    exchange: example
    encoding:
      codec: json
    properties:
      priority: 1
```

```yaml
sinks:
  sink_amqp:
    type: amqp
    inputs:
      - prioritize
    connection_string: amqp://user:password@127.0.0.1:5672/%2f?timeout=10
    exchange: example
    encoding:
      codec: json
    properties:
      priority: "{{ .priority }}"
```

pront commented on src/template.rs:
```diff
@@ -36,6 +36,38 @@ pub enum TemplateRenderingError {
+#[derive(Clone, Debug)]
+#[configurable_component]
+#[serde(untagged)]
+enum TemplateSource {
```
This is smart 👍 
Can we add some test coverage? 🙏

aramperes replied:
Done!

pront commented on src/sinks/amqp/config.rs:
```diff
@@ -33,7 +39,36 @@ impl AmqpPropertiesConfig {
+                    prop = prop.with_priority(priority);
+                }
+                Err(error) => {
+                    warn!(
```
Hm, I wonder if this is 'too spammy' because we can potentially have one per event.

aramperes replied:
I think this is why `internal_log_rate_limit` is set to true? Unless that doesn't throttle the logs

pront replied:
This is good, yes 👍 
We should probably make `internal_log_rate_limit` true by default.

pront commented:
> There could be a performance optimization done for numeric values to skip the `Number -> TemplateSource -> String -> Template -> String -> u8` cycle when evaluating the AMQP template, but that would require some more changes to the `Template` struct.

I think it's worth pursuing this. Also, I wonder if there's an optimization opportunity for `String`s by using `Cow`. But I don't think it's a blocker for this PR.

pront commented on changelog.d/add_priority_amqp_sink.enhancement.md:
```diff
@@ -0,0 +1,4 @@
+Support was added for `priority` on the `amqp` sink, to set a priority on messages sent.
+The priority value can be templated to an integer between 0 and 255 (inclusive).
```
Let's also mention that templates can now parse `i64`, `u64` and `f64` values?

pront commented:
> I think it's worth pursuing this. Also, I wonder if there's an optimization opportunity for `String`s by using `Cow`. But I don't think it's a blocker for this PR.

Is this something you want to add to this PR? (again optional)

aramperes commented:
> > I think it's worth pursuing this. Also, I wonder if there's an optimization opportunity for `String`s by using `Cow`. But I don't think it's a blocker for this PR.
> 
> Is this something you want to add to this PR? (again optional)

Hey. Unfortunately I don't have time this month to investigate this optimization. Might be better to merge and fix separately, I don't foresee a major performance loss here

pront commented on changelog.d/add_priority_amqp_sink.enhancement.md:
```diff
@@ -0,0 +1,4 @@
+Support was added for `priority` on the `amqp` sink, to set a priority on messages sent.
+The priority value can be templated to an integer between 0 and 255 (inclusive).
```
```suggestion
The `amqp` sink now supports setting the `priority` for messages. The priority value can be templated to an integer between 0 and 255 (inclusive). Also, this new field is template-able and Vector templates can now render numerical values.
```

aramperes replied:
Thank you!

pront commented:
Some clippy checks are failing, otherwise this is good to go 🚀

aramperes commented:
No worries, I'll get to it this weekend. Maybe some optimizations if I have time.

aramperes commented:
@pront Could you help me figure out what this build error means?

```
cargo build --quiet
target/debug/vector generate-schema > /tmp/vector-config-schema.json 2>/dev/null
make: *** [Makefile:679: generate-component-docs] Error 70
Error: Process completed with exit code 2.
```

aramperes commented:
After some more investigation, I found that the schema generator is not able to tell if the `Template` type is always a String because of the new `TemplateSource` sub-type. Furthermore `TemplateSource` could also be a Number so it would also be inaccurate to make it an `InstanceType::String`. This causes some schema validation to fail where `Template` could be used as a Map key before this change (allowing a Number as a template means you could have a Number as a Map key).

Maybe it would be cleaner to make separate types for `(String)Template` and `IntTemplate`?

pront commented:
> Maybe it would be cleaner to make separate types for `(String)Template` and `IntTemplate`?

Sounds reasonable, I will have to take another look at this whole PR.

pront commented on src/template.rs:
```diff
@@ -236,6 +244,205 @@ impl Template {
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+#[configurable_component]
+#[serde(untagged)]
+enum UIntTemplateSource {
```
Nit
```suggestion
enum UnsignedIntTemplateSource {
```

pront commented on src/template.rs:
```diff
@@ -236,6 +244,205 @@ impl Template {
+
+impl fmt::Display for UIntTemplateSource {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Number(i) => write!(f, "{}", i),
+            Self::String(s) => write!(f, "{}", s),
+        }
```
```suggestion
        match self {
            Self::Number(i) => i.fmt(f),
            Self::String(s) => s.fmt(f),
        }
```

pront reviewed (approved):
Nice!
